rhg_compute_tools/utils.py

```diff
@@ -437,3 +440,123 @@ def inner(*args, **kwargs):
+    pros and cons:
+
+    1. Using python's native ``threading`` module. If you are executing ``func`` outside
+       of a ``dask`` worker, you likely will want this approach. It may be slightly
+       faster and has the benefit of starting the timeout clock when the function starts
+       executing (rather than when the function is *submitted* to a dask scheduler).
+       **Note**: This approach will also work if calling ``func`` *from* a dask worker,
+       but only if the cluster was set up such that ``threads_per_worker=1``. Otherwise,
+       this may cause issues if used from a dask worker.
+    2. Using ``dask``. If you would like a dask worker to execute this function, you
+       likely will want this approach. It can be executed from a dask worker regardless
+       of the number of threads per worker (see above), but has the downside that the
+       timeout clock begins once ``func`` is submitted, rather than when it begins
+       executing.
+
+    Parameters
+    ----------
+    func : callable
+        The function you would like to execute with a timeout backoff.
+    retry_freq : float
+        The number of seconds to wait between successive retries of ``func``.
+    n_retries : int
+        The number of retries to attempt before raising an error if none were successful
+    use_dask : bool
+        If true, will try to use the ``dask``-based implementation (see description
+        above). If no ``Client`` instance is present, will fall back to
+        ``use_dask=False``.
+
+    Returns
+    -------
+    The return value of ``func``
+
+    Raises
+    ------
+    dask.distributed.TimeoutError :
+        If the function does not execute successfully in the specified ``retry_freq``,
+        after trying ``n_retries`` times.
+    ValueError :
+        If ``use_dask=True``, and a ``Client`` instance is present, but this fucntion is
+        executed from the client (rather than as a task submitted to a worker), you will
+        get ``ValueError("No workers found")``.
+
+    Examples
+    --------
+    .. code-block:: python
+
+        >>> import time
+        >>> @retry_with_timeout(retry_freq=.5, n_retries=1)
+        ... def wait_func(timeout):
+        ...     time.sleep(timeout)
+        ...     print("success")
+        >>> wait_func(.1)
+        success
+        >>> wait_func(1)
+        Traceback (most recent call last):
+            ...
+        asyncio.exceptions.TimeoutError: Func did not complete successfully in allowed time/number of retries.
+    """
+
+    # if use_dask specified, check if there is an active client, otherwise set to false
+    if use_dask:
+        try:
+            dd.get_client()
+        except ValueError:
+            use_dask = False
+
+    @functools.wraps(func)
+    def inner(*args, **kwargs):
+        if use_dask:
+            # dask version
+            with dd.worker_client() as client:
+                for retry_n in range(n_retries):
```

Comment:
nitpick: to me this seems like `n_tries` not `n_retries` :)

Reply:
haha great point. fixed